synapse_data_warehouse/synapse/dynamic_tables/V2.46.0__certifiedquiz_latest_dynamic_table.sql

```diff
@@ -0,0 +1,50 @@
+
+CREATE OR REPLACE DYNAMIC TABLE CERTIFIEDQUIZ_LATEST_DYNAMIC
+    TARGET_LAG = '1 day'
+    WAREHOUSE = compute_xsmall
+    COMMENT = 'This table, indexed by USER_ID, contains the latest snapshot of the certification quiz submissions by a Synapse user.'
+    (
+        CHANGE_TYPE VARCHAR(10) COMMENT 'The type of change to the record e.g., CREATE, UPDATE. CREATE indicates an user submitting a quiz results in a new submission of the quiz while UPDATE indicates a change to the pre-existing user record usually to indicate it being revoked. There are no DELETE changes.',
+        CHANGE_TIMESTAMP TIMESTAMP_NTZ(9) COMMENT 'The time when the change happened. For CREATE changes, it is when the user submitted the quiz. For UPDATE changes, it is when the pre-existing user record changed (e.g. got revoked)',
+        SNAPSHOT_TIMESTAMP TIMESTAMP_NTZ(9) COMMENT 'The time when the snapshot was taken (It is usually after the change happened).',
+        RESPONSE_ID NUMBER(38,0) COMMENT 'The unique identifier of a response wherein a user submitted a set of answers while participating in the quiz.',
+        USER_ID NUMBER(38,0) COMMENT 'The unique identifier of the user who submitted the quiz.',
+        PASSED BOOLEAN COMMENT 'If true, the user passed the quiz.',
+        PASSED_ON TIMESTAMP_NTZ(9) COMMENT 'The date on which the user submit the quiz, regardless of whether user passed or failed the test.',
+        STACK VARCHAR(10) COMMENT 'The stack (prod, dev) on which the quiz record was processed.',
+        INSTANCE VARCHAR(10) COMMENT 'The version of the stack that processed the quiz record.',
+        SNAPSHOT_DATE DATE COMMENT 'The snapshot_timestamp field is converted into a date and stored in the snapshot_date field.',
+        REVOKED BOOLEAN COMMENT 'If true, the record was revoked by an ACT member.',
+        REVOKED_ON TIMESTAMP_NTZ(9) COMMENT 'The date/time when the record was revoked, can be null if the record was never revoked.',
+        CERTIFIED BOOLEAN COMMENT 'If true the user is certified through this record, can be true if passed is true and revoked is false.'
+    )
+    AS
+        WITH latest_unique_rows AS (
+            SELECT
+                CHANGE_TYPE,
+                CHANGE_TIMESTAMP,
+                SNAPSHOT_TIMESTAMP,
+                RESPONSE_ID,
+                USER_ID,
+                PASSED,
+                PASSED_ON,
+                STACK,
+                INSTANCE,
+                SNAPSHOT_DATE,
+                REVOKED,
+                REVOKED_ON,
+                CERTIFIED
+            FROM
+                {{database_name}}.SYNAPSE_RAW.CERTIFIEDQUIZSNAPSHOTS --noqa: TMP
+            WHERE
+                SNAPSHOT_TIMESTAMP >= CURRENT_TIMESTAMP - INTERVAL '14 DAYS'
+            QUALIFY ROW_NUMBER() OVER (
+                    PARTITION BY USER_ID
+                    ORDER BY RESPONSE_ID DESC, CHANGE_TIMESTAMP DESC, SNAPSHOT_TIMESTAMP DESC
+                ) = 1
```

Comment:
Q1: Could it be possible for the same `USER_ID` to take the quiz multiple times, and if so, would we want to now about all those attempts?

Q2: Should this also take into account the `STACK` as well? It's a bit of an edge case specific to those with a dev/staging account, however, I wonder if this could cause us to drop a record from the prod stack if someone was doing testing in staging.

Reply:
Going to let @thomasyu888 and @philerooski chime in here for these to add any extra context

Q1: There was discussion earlier, and if I'm understanding it correctly, the conclusion of that discussion is that this table should just be the latest state table (so latest certified submission quiz state for a user) and we should have another de-duplicated table of all the latest submissions that a single user will take.

Q2: Great question, maybe I don't understand what `STACK` means then? We have `dev` and `prod` databases, I thought the `STACK` presents what database we are in so we'd have separate `certified quiz snapshots` tables for each database so there wouldn't be a snapshots table that has both `prod` and `dev` when we pull from it to create this table?

Reply:
> certified quiz snapshots tables for each database so there wouldn't be a snapshots table that has both prod and dev when we pull from it to create this table.

Great, if that's already filtered out - then there isn't a concern here

Reply:
I looked at the distinct values of `stack` and in SYNAPSE_DATA_WAREHOUSE this column is only ever "prod" and in SYNAPSE_DATA_WAREHOUSE_DEV it's only ever "dev".